File: gymgenie-mobile/iosApp/iosApp/Components/WorkoutPlanCard.swift
import SwiftUI
import Shared

/// Shared card representing a single workout plan.
///
/// Used in two contexts:
///  - The workouts catalog list, where the eye/start action pair is shown via
///    a non-nil `onView` handler.
///  - The home tab pager, where `onView` is `nil` and the card collapses to a
///    single full-width "Начать тренировку" CTA on a deep-ink background.
///
/// Active plans get a coral border + a 3pt top stripe so they read at a glance
/// without forcing the layout to grow taller. The body is wrapped in a
/// coral → white vertical gradient that signals the action color without
/// overpowering the content.
struct WorkoutPlanCard: View {
    let plan: WorkoutPlanShortResponse
    var onView: (() -> Void)? = nil
    var onStart: () -> Void = {}

    private var visual: MuscleGroupVisualStyle { muscleGroupStyle(plan.primaryMuscleGroup) }
    private var isAi: Bool { plan.createdBy.uppercased() == "AI" }
    private var isRecurring: Bool { plan.scheduleType.uppercased() == "RECURRING" }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if plan.isActive {
                Rectangle()
                    .fill(Palette.coral)
                    .frame(height: 3)
            }

            VStack(alignment: .leading, spacing: 0) {
                // Row 1: image + title + badge
                HStack(alignment: .center, spacing: 8) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 14)
                            .fill(visual.bg)
                        Text("🏋️").font(.system(size: 22))
                    }
                    .frame(width: 48, height: 48)

                    Text(plan.name)
                        .font(.system(size: 18, weight: .bold))
                        .foregroundColor(Palette.deepInk)
                        .lineLimit(2)
                        .frame(maxWidth: .infinity, alignment: .leading)

                    sourceBadge
                }

                Spacer().frame(height: 8)

                // Row 2: description (always reserves 2-line height)
                let descText: String = {
                    guard let d = plan.description_, !d.trimmingCharacters(in: .whitespaces).isEmpty else { return "\n" }
                    return d
                }()
                Text(descText)
                    .font(.system(size: 14))
                    .foregroundColor(descText == "\n" ? .clear : Color(red: 0.361, green: 0.361, blue: 0.388))
                    .lineLimit(2)
                    .frame(maxWidth: .infinity, alignment: .leading)

                Spacer().frame(height: 12)

                chipsRow

                Spacer().frame(height: 12)

                footer
            }
            .padding(16)
        }
        .background(
            LinearGradient(
                gradient: Gradient(colors: [Palette.coral.opacity(0.1), .white]),
                startPoint: .top,
                endPoint: .bottom
            )
        )
        .clipShape(RoundedRectangle(cornerRadius: 20))
        .overlay(
            RoundedRectangle(cornerRadius: 20)
                .strokeBorder(
                    plan.isActive ? Palette.coral : Color(red: 0.929, green: 0.929, blue: 0.937),
                    lineWidth: plan.isActive ? 2 : 1.5
                )
        )
    }

    // MARK: - Source badge

    private var sourceBadge: some View {
        Group {
            if isAi {
                Text("✦ AI")
                    .font(.system(size: 14, weight: .bold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(Capsule().fill(Palette.coral))
            } else {
                Text("Ручная")
                    .font(.system(size: 14, weight: .bold))
                    .foregroundColor(Color(red: 0.361, green: 0.361, blue: 0.388))
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(Capsule().fill(Color(red: 0.957, green: 0.957, blue: 0.965)))
            }
        }
    }

    // MARK: - Chips

    private var chipsRow: some View {
        HStack(spacing: 6) {
            InfoChip(icon: "clock", text: "~\(plan.estimatedMinutes) мин")
            InfoChip(icon: "list.bullet", text: "\(plan.exercisesCount) упр.")
            scheduleChip
        }
    }

    private var scheduleChip: some View {
        Group {
            if isRecurring {
                let days = formatScheduleDays(Array(plan.scheduleDays))
                HStack(spacing: 4) {
                    Image(systemName: "repeat").font(.system(size: 13, weight: .semibold))
                    Text(days).font(.system(size: 14, weight: .semibold))
                }
                .foregroundColor(Color(red: 0.914, green: 0.290, blue: 0.173))
                .padding(.horizontal, 10).padding(.vertical, 5)
                .background(RoundedRectangle(cornerRadius: 10).fill(Color(red: 1.0, green: 0.957, blue: 0.941)))
            } else {
                HStack(spacing: 4) {
                    Image(systemName: "clock").font(.system(size: 13, weight: .semibold))
                    Text("Разовая").font(.system(size: 14, weight: .semibold))
                }
                .foregroundColor(Color(red: 0.227, green: 0.227, blue: 0.251))
                .padding(.horizontal, 10).padding(.vertical, 5)
                .background(RoundedRectangle(cornerRadius: 10).fill(Color(red: 0.957, green: 0.957, blue: 0.965)))
            }
        }
    }

    // MARK: - Footer

    @ViewBuilder
    private var footer: some View {
        if let onView = onView {
            HStack(spacing: 10) {
                Button(action: onView) {
                    Image(systemName: "eye")
                        .font(.system(size: 16, weight: .medium))
                        .foregroundColor(Palette.deepInk)
                        .frame(width: 48, height: 48)
                        .background(
                            RoundedRectangle(cornerRadius: 14)
                                .fill(Color.white)
                        )
                        .overlay(
                            RoundedRectangle(cornerRadius: 14)
                                .strokeBorder(Color(red: 0.929, green: 0.929, blue: 0.937), lineWidth: 1.5)
                        )
                }
                .buttonStyle(.plain)

                Button(action: onStart) {
                    HStack(spacing: 6) {
                        Image(systemName: "play.fill").font(.system(size: 14, weight: .bold))
                        Text("Начать тренировку").font(.system(size: 18, weight: .bold))
                    }
                    .foregroundColor(.white)
                    .frame(maxWidth: .infinity)
                    .frame(height: 48)
                    .background(RoundedRectangle(cornerRadius: 14).fill(Palette.coral))
                }
                .buttonStyle(.plain)
            }
        } else {
            Button(action: onStart) {
                HStack(spacing: 8) {
                    Image(systemName: "play.fill").font(.system(size: 14, weight: .bold))
                    Text("Начать тренировку").font(.system(size: 18, weight: .bold))
                }
                .foregroundColor(.white)
                .frame(maxWidth: .infinity)
                .frame(height: 48)
                .background(RoundedRectangle(cornerRadius: 14).fill(Palette.deepInk))
            }
            .buttonStyle(.plain)
        }
    }
}

// MARK: - InfoChip

private struct InfoChip: View {
    let icon: String
    let text: String

    var body: some View {
        HStack(spacing: 4) {
            Image(systemName: icon).font(.system(size: 13, weight: .semibold))
            Text(text).font(.system(size: 14, weight: .semibold))
        }
        .foregroundColor(Color(red: 0.227, green: 0.227, blue: 0.251))
        .padding(.horizontal, 10).padding(.vertical, 5)
        .background(RoundedRectangle(cornerRadius: 10).fill(Color(red: 0.957, green: 0.957, blue: 0.965)))
    }
}

// MARK: - Muscle group visual style

private struct MuscleGroupVisualStyle {
    let bg: Color
    let emoji: String
}

private func muscleGroupStyle(_ group: String?) -> MuscleGroupVisualStyle {
    switch group?.uppercased() {
    case "CHEST":
        return .init(bg: Color(red: 1.0, green: 0.910, blue: 0.886), emoji: "🫁")
    case "BACK":
        return .init(bg: Color(red: 0.902, green: 0.933, blue: 1.0), emoji: "🦾")
    case "SHOULDERS", "SHOULDER":
        return .init(bg: Color(red: 0.902, green: 0.933, blue: 1.0), emoji: "💪")
    case "BICEPS", "TRICEPS", "FOREARMS", "ARMS":
        return .init(bg: Color(red: 1.0, green: 0.957, blue: 0.839), emoji: "💪")
    case "ABS", "CORE":
        return .init(bg: Color(red: 0.918, green: 0.902, blue: 1.0), emoji: "🔥")
    case "QUADRICEPS", "HAMSTRINGS", "GLUTES", "CALVES", "LEGS":
        return .init(bg: Color(red: 0.910, green: 0.969, blue: 0.910), emoji: "🦵")
    case "CARDIO":
        return .init(bg: Color(red: 0.988, green: 0.910, blue: 0.949), emoji: "❤️")
    case "FULL_BODY":
        return .init(bg: Color(red: 1.0, green: 0.957, blue: 0.839), emoji: "⭐")
    default:
        return .init(bg: Color(red: 1.0, green: 0.910, blue: 0.886), emoji: "🏋")
    }
}

// MARK: - Schedule day formatting

private let dayAbbreviations: [(String, String)] = [
    ("MONDAY", "Пн"),
    ("TUESDAY", "Вт"),
    ("WEDNESDAY", "Ср"),
    ("THURSDAY", "Чт"),
    ("FRIDAY", "Пт"),
    ("SATURDAY", "Сб"),
    ("SUNDAY", "Вс"),
]

private func formatScheduleDays(_ days: [String]) -> String {
    if days.isEmpty { return "Постоянная" }
    let upper = Set(days.map { $0.uppercased() })
    let ordered = dayAbbreviations.filter { upper.contains($0.0) }.map { $0.1 }
    return ordered.isEmpty ? "Постоянная" : ordered.joined(separator: " · ")
}

File: gymgenie-mobile/composeApp/src/androidMain/kotlin/com/asc/gymgenie/ui/components/WorkoutPlanCard.kt
package com.asc.gymgenie.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.List
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Repeat
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.asc.gymgenie.workout.WorkoutPlanShortResponse
import com.asc.gymgenie.workout.WorkoutScheduleType

private val Coral = Color(0xFFFF5A3C)
private val CoralDark = Color(0xFFE94A2C)
private val CoralTint = Color(0xFFFFF4F0)
private val BorderGray = Color(0xFFEDEDEF)
private val SoftGray = Color(0xFFF4F4F6)
private val ChipText = Color(0xFF3A3A40)
private val MutedManualText = Color(0xFF5C5C63)
private val InkBlack = Color(0xFF0A0A0A)
private val InkMuted = Color(0xFF8B8B92)

/**
 * Shared card representing a single workout plan.
 *
 * Used in two contexts:
 *  - The workouts catalog list, where the eye/start action pair is shown via a
 *    non-null [onView] handler.
 *  - The home tab pager, where [onView] is `null` and the card collapses to a
 *    single full-width "Начать тренировку" CTA on a deep-ink background.
 *
 * Active plans get a coral border + a 3dp top stripe so they read at a glance
 * without forcing the layout to grow taller. The body is wrapped in a coral →
 * white vertical gradient that signals the action color without overpowering
 * the content.
 */
@Composable
fun WorkoutPlanCard(
    plan: WorkoutPlanShortResponse,
    onView: (() -> Unit)? = null,
    onStart: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val borderColor = if (plan.isActive) Coral else BorderGray
    val borderWidth = if (plan.isActive) 2.dp else 1.5.dp
    val isAi = plan.createdBy.equals("AI", ignoreCase = true)
    val muscleColors = muscleGroupColors(plan.primaryMuscleGroup)
    val isRecurring = plan.scheduleType.equals(WorkoutScheduleType.RECURRING.name, ignoreCase = true)

    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(20.dp))
            .background(
                brush = Brush.verticalGradient(
                    colors = listOf(Coral.copy(alpha = 0.1f), Color.White),
                ),
            )
            .border(borderWidth, borderColor, RoundedCornerShape(20.dp)),
    ) {
        if (plan.isActive) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(3.dp)
                    .background(Coral),
            )
        }

        Column(modifier = Modifier.padding(horizontal = 16.dp, vertical = 16.dp)) {
            HeaderRow(
                plan = plan,
                muscleColors = muscleColors,
                isAi = isAi,
            )

            Spacer(modifier = Modifier.height(14.dp))

            ChipsRow(
                plan = plan,
                isRecurring = isRecurring,
            )

            Spacer(modifier = Modifier.height(14.dp))

            FooterActions(
                onView = onView,
                onStart = onStart,
            )
        }
    }
}

@Composable
private fun HeaderRow(
    plan: WorkoutPlanShortResponse,
    muscleColors: MuscleGroupColors,
    isAi: Boolean,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            modifier = Modifier
                .size(48.dp)
                .clip(RoundedCornerShape(14.dp))
                .background(muscleColors.background),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = muscleGroupCardEmoji(plan.primaryMuscleGroup),
                fontSize = 22.sp,
            )
        }

        Spacer(modifier = Modifier.width(12.dp))

        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = plan.name,
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
                color = InkBlack,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
            plan.description?.takeIf { it.isNotBlank() }?.let { desc ->
                Spacer(modifier = Modifier.height(2.dp))
                Text(
                    text = desc,
                    fontSize = 12.sp,
                    color = InkMuted,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }

        Spacer(modifier = Modifier.width(8.dp))

        SourceBadge(isAi = isAi)
    }
}

@Composable
private fun SourceBadge(isAi: Boolean) {
    if (isAi) {
        Text(
            text = "✦ AI",
            fontSize = 11.sp,
            fontWeight = FontWeight.Bold,
            color = Color.White,
            modifier = Modifier
                .clip(RoundedCornerShape(20.dp))
                .background(Coral)
                .padding(horizontal = 9.dp, vertical = 5.dp),
        )
    } else {
        Text(
            text = "Ручная",
            fontSize = 11.sp,
            fontWeight = FontWeight.Bold,
            color = MutedManualText,
            modifier = Modifier
                .clip(RoundedCornerShape(20.dp))
                .background(SoftGray)
                .padding(horizontal = 9.dp, vertical = 5.dp),
        )
    }
}

@Composable
private fun ChipsRow(
    plan: WorkoutPlanShortResponse,
    isRecurring: Boolean,
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        InfoChip(
            icon = Icons.Filled.Schedule,
            text = "~${estimatedMinutes(plan)} мин",
        )
        InfoChip(
            icon = Icons.AutoMirrored.Filled.List,
            text = "${plan.exercisesCount} упр.",
        )
        ScheduleChip(plan = plan, isRecurring = isRecurring)
    }
}

@Composable
private fun InfoChip(
    icon: ImageVector,
    text: String,
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .clip(RoundedCornerShape(10.dp))
            .background(SoftGray)
            .padding(horizontal = 10.dp, vertical = 5.dp),
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = ChipText,
            modifier = Modifier.size(13.dp),
        )
        Spacer(modifier = Modifier.width(4.dp))
        Text(
            text = text,
            fontSize = 12.sp,
            fontWeight = FontWeight.SemiBold,
            color = ChipText,
        )
    }
}

@Composable
private fun ScheduleChip(
    plan: WorkoutPlanShortResponse,
    isRecurring: Boolean,
) {
    val (background, foreground, label, icon) = if (isRecurring) {
        ScheduleChipStyle(
            background = CoralTint,
            foreground = CoralDark,
            label = formatRecurringDays(plan.scheduleDays),
            icon = Icons.Filled.Repeat,
        )
    } else {
        ScheduleChipStyle(
            background = SoftGray,
            foreground = ChipText,
            label = "Разовая",
            icon = Icons.Filled.Schedule,
        )
    }
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .clip(RoundedCornerShape(10.dp))
            .background(background)
            .padding(horizontal = 10.dp, vertical = 5.dp),
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = foreground,
            modifier = Modifier.size(13.dp),
        )
        Spacer(modifier = Modifier.width(4.dp))
        Text(
            text = label,
            fontSize = 12.sp,
            fontWeight = FontWeight.SemiBold,
            color = foreground,
        )
    }
}

@Composable
private fun FooterActions(
    onView: (() -> Unit)?,
    onStart: () -> Unit,
) {
    if (onView != null) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Box(
                modifier = Modifier
                    .size(48.dp)
                    .clip(RoundedCornerShape(14.dp))
                    .background(Color.White)
                    .border(1.5.dp, BorderGray, RoundedCornerShape(14.dp))
                    .clickable { onView() },
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = Icons.Filled.Visibility,
                    contentDescription = "Просмотр",
                    tint = InkBlack,
                    modifier = Modifier.size(20.dp),
                )
            }

            Row(
                modifier = Modifier
                    .weight(1f)
                    .height(48.dp)
                    .clip(RoundedCornerShape(14.dp))
                    .background(Coral)
                    .clickable { onStart() }
                    .padding(horizontal = 16.dp),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    imageVector = Icons.Filled.PlayArrow,
                    contentDescription = null,
                    tint = Color.White,
                    modifier = Modifier.size(18.dp),
                )
                Spacer(modifier = Modifier.width(6.dp))
                Text(
                    text = "Начать",
                    fontSize = 15.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color.White,
                )
            }
        }
    } else {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(48.dp)
                .clip(RoundedCornerShape(14.dp))
                .background(InkBlack)
                .clickable { onStart() }
                .padding(horizontal = 16.dp),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = Icons.Filled.PlayArrow,
                contentDescription = null,
                tint = Color.White,
                modifier = Modifier.size(16.dp),
            )
            Spacer(modifier = Modifier.width(8.dp))
            Text(
                text = "Начать тренировку",
                fontSize = 15.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White,
            )
        }
    }
}

private data class ScheduleChipStyle(
    val background: Color,
    val foreground: Color,
    val label: String,
    val icon: ImageVector,
)

internal data class MuscleGroupColors(
    val background: Color,
    val foreground: Color,
)

internal fun muscleGroupColors(group: String?): MuscleGroupColors {
    return when (group?.uppercase()) {
        "CHEST" -> MuscleGroupColors(Color(0xFFFFE8E2), Color(0xFFE94A2C))
        "BACK" -> MuscleGroupColors(Color(0xFFE6EEFF), Color(0xFF3B5BDB))
        "QUADRICEPS", "HAMSTRINGS", "GLUTES", "CALVES", "LEGS" ->
            MuscleGroupColors(Color(0xFFE8F7E8), Color(0xFF2F9E44))
        "BICEPS", "TRICEPS", "FOREARMS", "ARMS" ->
            MuscleGroupColors(Color(0xFFFFF4D6), Color(0xFFB8860B))
        "CARDIO" -> MuscleGroupColors(Color(0xFFFCE8F2), Color(0xFFC2255C))
        "ABS", "CORE" -> MuscleGroupColors(Color(0xFFEAE6FF), Color(0xFF6741D9))
        "SHOULDERS", "SHOULDER" -> MuscleGroupColors(Color(0xFFE6EEFF), Color(0xFF3B5BDB))
        else -> MuscleGroupColors(Color(0xFFFFE8E2), Color(0xFFE94A2C))
    }
}

internal fun muscleGroupCardEmoji(group: String?): String = when (group?.uppercase()) {
    "CHEST" -> "🫁"
    "BACK" -> "🦾"
    "SHOULDERS", "SHOULDER" -> "💪"
    "BICEPS", "TRICEPS", "FOREARMS", "ARMS" -> "💪"
    "ABS", "CORE" -> "🔥"
    "QUADRICEPS", "HAMSTRINGS", "GLUTES", "CALVES", "LEGS" -> "🦵"
    "CARDIO" -> "❤️"
    "FULL_BODY" -> "⭐"
    else -> "🏋"
}

/**
 * Lightweight time estimate that doesn't pretend to be precise: each set is
 * assumed to take ~1 min including the configured rest interval. This keeps
 * the chip useful even for plans that haven't been executed yet (so we can't
 * pull a real average from session history).
 */
private fun estimatedMinutes(plan: WorkoutPlanShortResponse): Int {
    if (plan.totalSets <= 0) return 0
    val perSetSeconds = 30 + plan.restSeconds.coerceAtLeast(0)
    val seconds = plan.totalSets * perSetSeconds
    return (seconds / 60).coerceAtLeast(1)
}

private val DayAbbreviations = listOf(
    "MONDAY" to "Пн",
    "TUESDAY" to "Вт",
    "WEDNESDAY" to "Ср",
    "THURSDAY" to "Чт",
    "FRIDAY" to "Пт",
    "SATURDAY" to "Сб",
    "SUNDAY" to "Вс",
)

internal fun formatRecurringDays(days: List<String>): String {
    if (days.isEmpty()) return "Постоянная"
    val normalized = days.map { it.uppercase() }.toSet()
    val ordered = DayAbbreviations.filter { it.first in normalized }.map { it.second }
    return if (ordered.isEmpty()) "Постоянная" else ordered.joinToString(" · ")
}
